### PR TITLE
feat: expand Listboxes only to max height

### PR DESCRIPTION
**Pull request by DanielS-Qlik**
smart-ex

### CODE REVIEW

File: src/components/ExpandButton.tsx [hunk 1]
```diff
@@ -15,6 +16,7 @@ export const ExpandButton = ({ onClick }: FoldedListboxProps) => (
       sx={{
         backgroundColor: '#F0F0F0',
         width: '100%',
+        height: `${BUTTON_HEIGHT}px`,
```
Comment (johanlahti) [inline src/components/ExpandButton.tsx:19]: `px` is assumed when you send in an integer directly so can use:
`height: BUTTON_HEIGHT,`

File: src/components/FoldedListbox.tsx [hunk 2]
```diff
@@ -12,14 +12,12 @@ export const FoldedListbox = ({ layout }: FoldedListboxProps) => {
...
-        border="1px solid lightgrey"
-        padding="8px"
-      >
-        {fieldName}
-      </Typography>
-    </>
+    <Typography
+      variant="h6"
+      border="1px solid lightgrey"
+      padding="8px"
```
Comment (johanlahti) [inline src/components/FoldedListbox.tsx:18]: And for this I think `padding={1}` is equivalent of `8px`: https://mui.com/system/getting-started/the-sx-prop/#spacing – don't need to change, just mentioning.

File: src/components/ListboxGrid/ListboxGrid.tsx [hunk 3]
```diff
@@ -57,60 +69,40 @@ export default function ListboxGrid(props: ListboxGridProps) {
...
-                <ExpandButton onClick={handleOnFullscreen} />
-              </Grid>}
-              {resourcesFirstFolded.map((resource: IListboxResource) => (
-                <Grid item key={resource.id} width='100%' xs={1}>
-                  <FoldedListbox layout={resource.layout} />
-                </Grid>
-              ))}
-            </Grid>
+                {!!column?.items?.length && column.items.map((item: IListboxResource) => (
+                  <ColumnItem key={item.id} height={item.expand ? item.height : `${COLLAPSED_HEIGHT}px`}>
```
Comment (johanlahti) [inline src/components/ListboxGrid/ListboxGrid.tsx:86]: unnecessary `px`

File: src/components/ListboxGrid/distribute-resources.ts [hunk 4]
```diff
@@ -1,61 +1,232 @@
...
+
+    // Default case with zoom enabled, now overflow and (...) to zoom object
+    itemCount = Math.min(size.dimensionCount - usedCount, maxPerColumn);
+
+    // TODO:
+    // if (ResponsiveState.isSmallDevice) {
+    // On small device all items is in a single column and overflow scrolled
+    // itemCount = size.dimensionCount - usedCount;
+    // } else if (!zoomEnabled) {
+    if (!zoomEnabled) {
```
Comment (johanlahti) [inline src/components/ListboxGrid/distribute-resources.ts:78]: Are these and other TODO:s to be cleaned up in this PR?

File: src/components/ListboxGrid/distribute-resources.ts [hunk 5]
```diff
@@ -1,61 +1,232 @@
...
+      // but don't add more items then we have left
+      itemCount = Math.min(size.dimensionCount - usedCount, itemCount);
+    }
+
+    columns.push({
+      expand: false,
+      itemCount,
+    });
+
+    // Last columns and all items wont fit, show the there is more items not rendered cue
```
Comment (johanlahti) [inline src/components/ListboxGrid/distribute-resources.ts:93]: Not understanding the comment 😆

File: src/components/ListboxGrid/distribute-resources.ts [hunk 6]
```diff
@@ -1,61 +1,232 @@
...
+      column?.items?.forEach((item) => {
+        item.cardinal = getDimensionCardinal(item);
+      });
+
+      const sortedItems = column?.items?.concat().sort((a, b) => a.cardinal - b.cardinal);
+
+      if ((sortedItems?.length ?? 0) > 1) {
+        while (leftOverHeight > EXPANDED_HEIGHT) {
+          const expHeight = expandOne(sortedItems, leftOverHeight);
+          if (expHeight) {
```
Comment (johanlahti) [inline src/components/ListboxGrid/distribute-resources.ts:185]: What happens, and should happen, if `expHeight === 0`?

File: src/components/ListboxGrid/distribute-resources.ts [hunk 7]
```diff
@@ -1,61 +1,232 @@
...
+    // TODO:
+    // if (ResponsiveState.isSmallDevice) {
+    // On small device all items is in a single column and overflow scrolled
+    // itemCount = size.dimensionCount - usedCount;
+    // } else if (!zoomEnabled) {
+    if (!zoomEnabled) {
+      // Zoom not enabled (mashup), items evenly distributed and overflow scrolled
+      // distribute the remaining items evenly
+      itemCount = Math.ceil((size.dimensionCount - usedCount) / (maxColumns - columns.length));
+      // to make sure all items in the columns are visible QB-2993
```
Comment (johanlahti) [inline src/components/ListboxGrid/distribute-resources.ts:82]: Are these comments relevant still?
Any way we can avoid re-assigning `itemCount` by creating new, and more descriptive variables like:
```
const initialItemCount = …
const maxItemCount = …
const itemCount = …
```
Then we could get rid of the comments as well.

File: src/components/ListboxGrid/distribute-resources.ts [hunk 8]
```diff
@@ -1,61 +1,232 @@
...
+  const expanded = columns.filter((column) => column.expand);
+  const collapsed = columns.filter((column) => column.expand === false);
+
+  // If there is room to expand one item in the last column of collapsed items we expand it.
+  // If not we distribute the items as evenly as possible among the columns with collapsed items.
+  if (
+    collapsed.length > 0
+    && !collapsed[collapsed.length - 1].showAll
+    && haveRoomToExpandOne(size, collapsed[collapsed.length - 1])
+  ) {
```
Comment (johanlahti) [inline src/components/ListboxGrid/distribute-resources.ts:121]: Here also, I would prefer removing/shorten the comments and adding a descriptive variable:
```
const canExpand = collapsed.length > 0 && !collapsed[collapsed.length - 1].showAll && haveRoomToExpandOne(size, collapsed[collapsed.length - 1]);
if (canExpand) { …
```

File: src/components/ListboxGrid/distribute-resources.ts (hunk 4 above)
Comment (DanielS-Qlik) [reply src/components/ListboxGrid/distribute-resources.ts:78]: I need to test this with sense-client. I don't want to remove them before. I've added a 'responsive'-task to our board for this.

File: src/components/ListboxGrid/distribute-resources.ts (hunk 5 above)
Comment (DanielS-Qlik) [reply src/components/ListboxGrid/distribute-resources.ts:93]: I will refactor the comment :)

File: src/components/ListboxGrid/distribute-resources.ts (hunk 6 above)
Comment (DanielS-Qlik) [reply src/components/ListboxGrid/distribute-resources.ts:185]: Then the item is not expanded, it should go to `else`

**Pull request merged**